File: binary_search/1_binary_search/main.py
from typing import List 

def recursive_binary_search(l:int, r:int, nums:List[int], target:int) -> int:
    if l > r:
        return - 1
    m = l + (r-l) //2

    if nums[m] == target:
        return m 
    if nums[m] < target:
        return recursive_binary_search(m +1, r, nums,target)
    return recursive_binary_search(l, m -1, nums, target)

def search(nums:List[int], target: int) -> int:
    return recursive_binary_search(0, len(nums) - 1, nums, target)


def iterative_binary_search(nums:List[int], target: int) -> int:
    l, r = 0, len(nums) - 1
    while l < r:
        m = l + r ((r - l) // 2)
        if nums[m] > target:
            r = m -1 
        elif nums[m] < target:
            l = m +1 
        else:
            return m 
    return -1 

def upper_bound(nums:List[int], target:int) -> int: 
    l, r = 0, len(nums)
    while l < r:
        m = l + ((r - l) // 2)
        if nums[m] > target:
            r = m
        elif nums[m] <= target:
            l = m + 1 
    return l -1 if (l and nums[l - 1] == target) else -1


def lower_bound(nums:List[int], target:int) -> int:
    l, r = 0, len(nums)
    while l < r:
        m = l + ((r - l) // 2)
        if nums[m] >= target:
            r = m 
        elif nums[m] < target:
            l = m + 1
    return l if (l < len(nums) and nums[l] == target) else -1

def built_in(nums:List[int], target:int) -> int:
    import biset 
    index = biset.biset_left(nums, target)
    return index if index < len(nums) and nums[index] == target else - 1

File: binary_search/1_binary_search/Main.java
import java.util.Arrays;

public class Main {
    
    public int recursive_binary_search(int l, int r, int[] nums, int target){
        if (l > r) return -1;
        int m = l + (r - l) / 2;

        if (nums[m] == target) return m;
        return (nums[m] < target) ? 
        recursive_binary_search(m +1, r, nums, target):
        recursive_binary_search(l, m-1, nums, target);
    }
    public int search(int[] nums, int target){
        return recursive_binary_search(0, nums.length -1 , nums, target);
    }


    public int iterative_binary_search(int[] nums, int target){
        int l = 0, r=nums.length -1;
        while (l <= r){
            int m = l +((r - l) / 2);
            if (nums[m] > target){
                r = m -1;
            }else if (nums[m] < target){
                l = m +1;
            }else{
                return m;
            }
        }
        return -1;
    }

    public int upper_bound(int[] nums, int target){
        int l = 0, r = nums.length;

        while (l < r){
            int m = l + ((r - l)/ 2);
            if (nums[m] > target){
                r = m;
            }else{
                l = m +1;
            }
        }
        return (l > 0 && nums[l - 1] == target) ? l - 1 : -1;
    }
    public int lower_bound(int[] nums, int target){
        int l = 0, r = nums.length;
        while (l < r){
            int m = l + (r - l) / 2;
            if (nums[m] >= target){
                r = m;
            }else{
                l = m + 1;
            }
        }
        return (l < nums.length && nums[l] == target) ? l : -1;
    }

    public int built_in(int[] nums, int target){
        int index = Arrays.binarySearch(nums, target);
        return index >= 0 ? index : -1;
    }
}
